proposed/activity-improvements.md

```diff
@@ -0,0 +1,465 @@
+  - Allow attaching any object to the Activity object through the `Set/GetCustomProperty` for the sake of the extendability without manually sub-classing Activity or maintaining a dictionary mapping activity to the other objects. Examples of that, OT implementation still depends on Activity and maintains a table mapping from Activity to Span. Another example is some teams were working around the lack of missing Links and Contexts features and manually doing extra work to include these features. Although we are adding the Links and Contexts, we can get more future scenarios to extend Activity (e.g. attaching the status codes/exceptions).
+
+- The current usage pattern of the Activity is not neat. here is an example of the current usage:
+
+```C#
+var activityListener = new DiagnosticSource("Azure.Core.Http");
+
+// Outer check to see if anyone is subscribed
+if (activityListener.IsEnabled())
+{
+    Activity activity = null;
+    // Check if anyone cares about activity
+    if (activityListener.IsEnabled("Azure.Core.Http.Request"))
+    {
+        activity = new Activity("Azure.Core.Http.Request");
+        activity.AddTag..
+        activityListener.StartActivity(activity); // this does string concat and allocates every time
+    }
+
+    ...
+
+    if (activity != null)
+    {
+        activityListener.StopActivity(activity);
+    }
+}
+```
+
+Note in the pattern, the code trying to avoid creating the Activity object if there are no listeners interested to listen to such Activity creation. Also, the code later has to check if the activity object is created and then stop it.
+We are simplifying the pattern to something like:
+
+```C#
+    static ActivitySource source =  new ActivitySource("Azure.Core.Http", "4.0.0.0");
+    ...
+
+    // StartActivity can return null if no listener is enabled. Activity now is Disposable too which automatically stop the activity if it is created and started.
+    using (Activity? activity = source.StartActivity("Request"))
+    {
+        activity?.AddTag..
+    }
+```
+
+- Last, we are introducing ActivitySource type which can be used to efficiently listening to Activities created from specific component/sources or all sources. It is now reasonable for a listener such as OpenTelemetry to observe Activity instrumentation from any code component in the app by the `ActivitySource.AddActivityListener()`. Previously OpenTelemetry would have needed appropriate  knowledge of the strings that would be used for DiagnosticSource name and IsEnabled() callbacks for each Activity it wanted to receive callbacks for. When trying to create a new Activity object through the new introduced factory methods `ActivitySource.StartActivity`, the listener will have the opportunity to get a callback to decide if interested in having the Activity object get created or it can be sampled out to avoid the object creation if nobody else is listening or other listeners are not interested in such activity too.
```

Comment:
```suggestion
- Last, we are introducing ActivitySource type which can be used to efficiently listening to Activities created from specific component/sources or all sources. It is now reasonable for a listener such as OpenTelemetry to observe Activity instrumentation from any code component in the app by the `ActivitySource.AddActivityListener()`. Previously OpenTelemetry would have needed appropriate knowledge of the strings that would be used for `DiagnosticSource` name and `IsEnabled()` callbacks for each Activity it wanted to receive callbacks for. When trying to create a new Activity object through the new introduced factory methods `ActivitySource.StartActivity`, the listener will have the opportunity to get a callback to decide if interested in having the Activity object get created or it can be sampled out to avoid the object creation if nobody else is listening or other listeners are not interested in such activity too.
```